DOCSUP-8324: Document the support of replicas and shards in MySQL/PostgreSQL (#25688)
ClickHouse/ClickHouse: docs/en/engines/table-engines/integrations/ExternalDistributed.md

---
toc_priority: 12
toc_title: ExternalDistributed
---

# ExternalDistributed {#externaldistributed}

> The `ExternalDistributed` engine allows to perform `SELECT` and `INSERT` queries on data that is stored on a remote servers MySQL or PostgreSQL. Accepts [MySQL](../../../engines/table-engines/integrations/mysql.md) or [PostgreSQL](../../../engines/table-engines/integrations/postgresql.md) engines as an argument so sharding is possible.

Reviewer: No, the table is read-only.

-   Column types may differ from those in the original table. ClickHouse tries to [cast](../../../sql-reference/functions/type-conversion-functions.md#type_conversion_function-cast) values to the ClickHouse data types.
-   The `external_table_functions_use_nulls` setting defines how to handle Nullable columns. Default value: 1. If 0, the table function will not make nullable columns and will insert default values instead of nulls. This is also applicable for null values inside array data types.

**Engine Parameters**

-   `engine` — The table engine `MySQL` or `PostgreSQL`.
-   `host:port` — MySQL or PostgreSQL server address.
-   `database` — Remote database name.
-   `table` — Remote table name.
-   `user` — User name.
-   `password` — User password.

## Implementation Details {#implementation-details}
	
Supports multiple replicas that must be listed by `|` and shards must be listed by `,`. For example:

```sql
CREATE TABLE test_shards (id UInt32, name String, age UInt32, money UInt32) ENGINE = ExternalDistributed('MySQL', `mysql{1|2}:3306,mysql{3|4}:3306`, 'clickhouse', 'test_replicas', 'root', 'clickhouse');
```

> When specifying replicas, one of the available replicas will be selected for each of the shards when reading. If the connection failed, the next replica will be selected, and so on for all the replicas. If the connection attempt failed for all the replicas, the attempt will be repeated the same way several times.

Reviewer 2: ```suggestion
When specifying replicas, one of the available replicas is selected for each of the shards when reading. If the connection fails, the next replica is selected, and so on for all the replicas. If the connection attempt fails for all the replicas, the attempt is repeated the same way several times.
```

> You can specify up to any number of shards and to any number of replicas for each shard.

Reviewer 2: ```suggestion
You can specify any number of shards and number of replicas for each shard.
```

    ...
) ENGINE = ExternalDistributed('engine', 'host:port', 'database', 'table', 'user', 'password'[, `schema`]);
```

See a detailed description of the [CREATE TABLE](../../../sql-reference/statements/create/table.md#create-table-query) query.

The table structure can differ from the original table structure:

-   Column names should be the same as in the original table, but you can use just some of these columns and in any order.
-   Column types may differ from those in the original table. ClickHouse tries to [cast](../../../sql-reference/functions/type-conversion-functions.md#type_conversion_function-cast) values to the ClickHouse data types.
-   The `external_table_functions_use_nulls` setting defines how to handle Nullable columns. Default is 1, if 0 - table function will not make nullable columns and will insert default values instead of nulls. This is also applicable for null values inside array data types.

**Engine Parameters**

-   `engine` — The table engine: `MySQL` or `PostgreSQL`.
-   `host:port` — MySQL or PostgreSQL server address.
-   `database` — Remote database name.
-   `table` — Remote table name.
-   `user` — User name.
-   `password` — User password.

> -   `schema` — Non-default table schema. Optional.

Reviewer: `schema` parameter was not added, but should have been though.

## Implementation Details {#implementation-details}
	

> Supports multiple replicas that must be listed by a character `|`. For example:

Reviewer: And shards are listed with `,`.

```sql
CREATE TABLE test_shards (id UInt32, name String, age UInt32, money UInt32) ENGINE = ExternalDistributed('MySQL', `mysql{1|2}:3306,mysql{3|4}:3306`, 'clickhouse', 'test_replicas', 'root', 'clickhouse');
```

> When specifying replicas, one of the available replicas will be selected for each of the shards when reading. You can configure the algorithm for load balancing (the preference for which replica to access) – see the [load_balancing](../../../operations/settings/settings.md#settings-load_balancing) setting. If the connection with the server is not established, there will be an attempt to connect with a short timeout. If the connection failed, the next replica will be selected, and so on for all the replicas. If the connection attempt failed for all the replicas, the attempt will be repeated the same way several times. This works in favor of resiliency, but does not provide complete fault tolerance: a remote server might accept the connection, but might not work, or work poorly.

Reviewer: > You can configure the algorithm for load balancing (the preference for which replica to access) – see the [load_balancing]

No, it is not configurable via `load_balancing` setting.

> there will be an attempt to connect with a short timeout.

There is not timeout.

>  a remote server might accept the connection, but might not work

Questionable - if connection is accepted, then it is actually connected, but might disconnect any moment after.

> You can specify just one of the shards (in this case, query processing should be called remote, rather than distributed) or up to any number of shards. In each shard, you can specify from one to any number of replicas. You can specify a different number of replicas for each shard.

Reviewer: > You can specify just one of the shards 

In this case there is no point to use this table engine as it will be the same as using just MySQL/PostgreSQL storages directly.

> Each shard can have a weight defined in the config file. By default, the weight is equal to one. Data is distributed across shards in the amount proportional to the shard weight. For example, if there are two shards and the first has a weight of 9 while the second has a weight of 10, the first will be sent 9 / 19 parts of the rows, and the second will be sent 10 / 19.

Reviewer: > Each shard can have a weight defined in the config file.

Not true - there are no shard weights.

> To select the shard that a row of data is sent to, the sharding expression is analyzed, and its remainder is taken from dividing it by the total weight of the shards. The row is sent to the shard that corresponds to the half-interval of the remainders from `prev_weight` to `prev_weights + weight`, where `prev_weights` is the total weight of the shards with the smallest number, and `weight` is the weight of this shard. For example, if there are two shards, and the first has a weight of 9 while the second has a weight of 10, the row will be sent to the first shard for the remainders from the range \[0, 9), and to the second for the remainders from the range \[9, 19).

Reviewer: This line and below about shard weights cannot be applied as there are no shard weights.
